### PR TITLE
Add retry mechanism

### PR DESCRIPTION
**Pull request by altjohndev**
This PR overhauls the client layer by replacing [HTTPoison](https://hexdocs.pm/httpoison) with [Tesla](https://hexdocs.pm/tesla).

The retry mechanism is performed through [Tesla.Middleware.Retry](https://hexdocs.pm/tesla/Tesla.Middleware.Retry.html).

It also removes the `Agent` definition in favor of tuning through application configuration and function options.

This will produce a `MAJOR` release.

🎗️  [CHANGELOG](https://github.com/viodotcom/analytics-elixir/blob/9163b484ad14f932b737c967bf02160fcc3fc982/CHANGELOG.md#unreleased).

### CODE REVIEW

File: lib/segment/analytics/http.ex [hunk 1]
```diff
@@ -1,21 +1,73 @@
...
+    [
+      delay: config.retry_base_delay,
+      jitter_factor: config.retry_jitter_factor,
+      max_delay: config.retry_max_delay,
+      max_retries: config.max_retries,
+      should_retry: &should_retry?/1
+    ]
+  end
+
+  defp should_retry?({:ok, %Tesla.Env{} = env}) when response_ok?(env), do: false
```
Comment (prodis) [inline lib/segment/analytics/http.ex:54]: Missing typespec.

File: lib/segment/analytics/http.ex [hunk 2]
```diff
@@ -1,21 +1,73 @@
...
+  defp should_retry?(_result), do: true
+
+  defp meta_logger(%Config{} = config) do
```
Comment (prodis) [inline lib/segment/analytics/http.ex:58]: Missing typespec.

File: lib/segment.ex [hunk 3]
```diff
@@ -1,77 +1,85 @@
...
-  use Agent
+  @moduledoc """
+  Client for Segment API.
 
-  @type status :: :ok | :error
-
-  @default_endpoint "https://api.segment.io/v1/"
-
-  @spec start_link(String.t(), String.t()) :: {Segment.status(), pid}
-  def start_link(key, endpoint \\ @default_endpoint) do
```
Comment (prodis) [inline lib/segment.ex:9]: 🎗️ https://github.com/viodotcom/analytics-elixir/blob/9163b484ad14f932b737c967bf02160fcc3fc982/README.md#usage

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,21 +7,44 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
...
+### Changed
+
+- Replaced `HTTPoison` library with `Tesla`.
+
+### Removed
+
+- Removed `Agent` strategy in favor of configuration. See `t:Segment.options/0` for configuration
+  instructions.
+
+### Added
```
Comment (prodis) [inline CHANGELOG.md:19]: Please update README with some usage information about what was added, at least linking the modules/functions documentation.

**Review by prodis — APPROVED** [review]

**Review by dmarasquin — APPROVED** [review]

File: lib/segment.ex [hunk 5]
```diff
@@ -1,77 +1,85 @@
...
+  @typedoc """
+  Options to customize the operation.
+
+  It is possible to define options through application environment:
+
+      # in config.exs
+      import Config
+
+      config :segment,
+        disable_meta_logger: #{inspect(@default_config.drop_nil_fields)},
```
Comment (devjoaoGustavo) [inline lib/segment.ex]:
```suggestion
        disable_meta_logger: #{inspect(@default_config.disable_meta_logger)},
```

**Review by devjoaoGustavo — APPROVED** [review]

**Pull request merged**